File: python/11_Container_With_Most_Water.py
class Container_With_Most_Water:
  def maxArea(self, height: list[int]) -> int:
    l = 0
    r = len(height) - 1
    maxArea = 0
    maxHeight = max(height)
    while l < r:
      area = min(height[r],height[l]) * (r - l)
      maxArea = max(maxArea,area)
      if maxHeight * (r - l) <= maxArea:
        break
      if height[l] > height[r]:
        r -= 1
      else:
        l += 1
    return maxArea

File: c++/11_Container_With_Most_Water.cpp
#include <vector>

using namespace std;

class Container_With_Most_Water {
public:
  int maxArea(vector<int>& height) {
    int l = 0;
    int r = height.size() - 1;
    int maxArea = 0;
    while (l < r) {
      maxArea = max(maxArea, (min(height[l], height[r]) * (r - l)));
      height[r] < height[l] ? r-- : l++;
    }
    return maxArea;
  }
};